File: knowledge-bases/external-knowledge-bases/external-knowledge-bases-node/src/schemas.ts
import { z } from '@hono/zod-openapi';

/**
 * Placeholder for search filter schema.
 * Extend this based on your specific filter requirements.
 */
export const SearchFilterSchema = z
  .record(z.unknown())
  .openapi('SearchFilter', {
    description: 'The metadata filter to apply to the search',
  });

/**
 * Additional search options
 */
export const SearchOptionsSchema = z
  .object({
    include_vectors: z.boolean().optional().openapi({
      description: 'Whether to include the vector in the chunk',
    }),
    include_metadata: z.boolean().optional().openapi({
      description: 'Whether to include the metadata in the chunk',
    }),
    include_scores: z.boolean().optional().openapi({
      description: 'Whether to include the scores in the chunk',
    }),
  })
  .openapi('SearchOptions');

/**
 * Override the rerank configuration for this search.
 * If not provided, will use the knowledge base configured rerank settings.
 */
export const RerankConfigSchema = z
  .object({
    model: z.string().openapi({
      description:
        'The name of the rerank model to use. Refer to the model list (https://docs.orq.ai/docs/proxy#/rerank-models).',
      example: 'cohere/rerank-multilingual-v3.0',
    }),
    threshold: z
      .number()
      .min(0)
      .max(1)
      .default(0)
      .optional()
      .openapi({
        description:
          'The threshold value used to filter the rerank results, only documents with a relevance score greater than the threshold will be returned',
      }),
    top_k: z
      .number()
      .int()
      .min(1)
      .max(100)
      .default(10)
      .optional()
      .openapi({
        description:
          'The number of top results to return after reranking. If not provided, will default to the knowledge base configured `top_k`.',
      }),
  })
  .openapi('RerankConfig');

/**
 * Override the agentic RAG configuration for this search.
 * If not provided, will use the knowledge base configured agentic RAG settings.
 */
export const AgenticRagConfigSchema = z
  .object({
    model: z.string().openapi({
      description:
        'The name of the model for the Agent to use. Refer to the model list (https://docs.orq.ai/docs/proxy#/chat-models).',
    }),
  })
  .openapi('AgenticRagConfig');

/**
 * Schema for searching a knowledge base
 */
export const SearchKnowledgeRequestSchema = z
  .object({
    query: z.string().openapi({
      description: 'The query to use to search the knowledge base',
    }),
    top_k: z
      .number()
      .int()
      .min(1)
      .max(100)
      .optional()
      .openapi({
        description:
          'The number of results to return. If not provided, will default to the knowledge base configured `top_k`.',
      }),
    threshold: z
      .number()
      .min(0)
      .max(1)
      .optional()
      .openapi({
        description:
          'The threshold to apply to the search. If not provided, will default to the knowledge base configured `threshold`',
      }),
    search_type: z
      .enum(['vector_search', 'keyword_search', 'hybrid_search'])
      .default('hybrid_search')
      .optional()
      .openapi({
        description:
          'The type of search to perform. If not provided, will default to the knowledge base configured `retrieval_type`',
      }),
    filter_by: SearchFilterSchema.optional().openapi({
      description:
        'The metadata filter to apply to the search. Check the Searching a Knowledge Base documentation for more information.',
    }),
    search_options: SearchOptionsSchema.optional().openapi({
      description: 'Additional search options',
    }),
    rerank_config: RerankConfigSchema.optional().openapi({
      description:
        'Override the rerank configuration for this search. If not provided, will use the knowledge base configured rerank settings.',
    }),
    agentic_rag_config: AgenticRagConfigSchema.optional().openapi({
      description:
        'Override the agentic RAG configuration for this search. If not provided, will use the knowledge base configured agentic RAG settings.',
    }),
  })
  .openapi('SearchKnowledgeRequest');

/**
 * Scores for a search result
 */
export const ScoresSchema = z
  .object({
    rerank_score: z.number().optional(),
    search_score: z.number().optional(),
  })
  .openapi('Scores');

/**
 * A single match from the knowledge base search
 */
export const MatchSchema = z
  .object({
    id: z.string(),
    text: z.string(),
    vector: z.array(z.number()).optional(),
    metadata: z.record(z.string(), z.unknown()).optional(),
    scores: ScoresSchema.optional(),
  })
  .openapi('Match');

/**
 * Response schema for knowledge base search
 */
export const SearchKnowledgeResponseSchema = z
  .object({
    matches: z.array(MatchSchema),
  })
  .openapi('SearchKnowledgeResponse');

File: knowledge-bases/external-knowledge-bases/external-knowledge-bases-python/schemas.py
from typing import Any, Dict, List, Literal, Optional
from pydantic import BaseModel, Field, ConfigDict


class SearchFilterSchema(BaseModel):
    """
    Placeholder for search filter schema.
    Extend this based on your specific filter requirements.
    """
    model_config = ConfigDict(extra='allow')


class SearchOptionsSchema(BaseModel):
    """Additional search options"""
    include_vectors: Optional[bool] = Field(
        None,
        description="Whether to include the vector in the chunk"
    )
    include_metadata: Optional[bool] = Field(
        None,
        description="Whether to include the metadata in the chunk"
    )
    include_scores: Optional[bool] = Field(
        None,
        description="Whether to include the scores in the chunk"
    )


class RerankConfigSchema(BaseModel):
    """Override the rerank configuration for this search. If not provided, will use the knowledge base configured rerank settings."""
    model: str = Field(
        ...,
        description="The name of the rerank model to use. Refer to the model list (https://docs.orq.ai/docs/proxy#/rerank-models).",
        examples=["cohere/rerank-multilingual-v3.0"]
    )
    threshold: Optional[float] = Field(
        0,
        ge=0,
        le=1,
        description="The threshold value used to filter the rerank results, only documents with a relevance score greater than the threshold will be returned"
    )
    top_k: Optional[int] = Field(
        10,
        ge=1,
        le=100,
        description="The number of top results to return after reranking. If not provided, will default to the knowledge base configured `top_k`."
    )


class AgenticRagConfigSchema(BaseModel):
    """Override the agentic RAG configuration for this search. If not provided, will use the knowledge base configured agentic RAG settings."""
    model: str = Field(
        ...,
        description="The name of the model for the Agent to use. Refer to the model list (https://docs.orq.ai/docs/proxy#/chat-models)."
    )


class SearchKnowledgeRequestSchema(BaseModel):
    """Schema for searching a knowledge base"""
    query: str = Field(
        ...,
        description="The query to use to search the knowledge base"
    )
    top_k: Optional[int] = Field(
        None,
        ge=1,
        le=100,
        description="The number of results to return. If not provided, will default to the knowledge base configured `top_k`."
    )
    threshold: Optional[float] = Field(
        None,
        ge=0,
        le=1,
        description="The threshold to apply to the search. If not provided, will default to the knowledge base configured `threshold`"
    )
    search_type: Optional[Literal['vector_search', 'keyword_search', 'hybrid_search']] = Field(
        'hybrid_search',
        description="The type of search to perform. If not provided, will default to the knowledge base configured `retrieval_type`"
    )
    filter_by: Optional[SearchFilterSchema] = Field(
        None,
        description="The metadata filter to apply to the search. Check the Searching a Knowledge Base documentation for more information."
    )
    search_options: Optional[SearchOptionsSchema] = Field(
        None,
        description="Additional search options"
    )
    rerank_config: Optional[RerankConfigSchema] = Field(
        None,
        description="Override the rerank configuration for this search. If not provided, will use the knowledge base configured rerank settings."
    )
    agentic_rag_config: Optional[AgenticRagConfigSchema] = Field(
        None,
        description="Override the agentic RAG configuration for this search. If not provided, will use the knowledge base configured agentic RAG settings."
    )


class ScoresSchema(BaseModel):
    """Scores for a search result"""
    rerank_score: Optional[float] = None
    search_score: Optional[float] = None


class MatchSchema(BaseModel):
    """A single match from the knowledge base search"""
    id: str
    text: str
    vector: Optional[List[float]] = None
    metadata: Optional[Dict[str, Any]] = None
    scores: Optional[ScoresSchema] = None


class SearchKnowledgeResponseSchema(BaseModel):
    """Response schema for knowledge base search"""
    matches: List[MatchSchema]
